File: libaums/src/main/java/com/github/mjdev/libaums/driver/file/FileBlockDeviceDriver.java
package com.github.mjdev.libaums.driver.file;

import com.github.mjdev.libaums.driver.BlockDeviceDriver;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;

/**
 * Created by magnusja on 01/08/17.
 */

public class FileBlockDeviceDriver implements BlockDeviceDriver {
    private RandomAccessFile file;
    private int blockSize;
    private int byteOffset;

    public FileBlockDeviceDriver(File file, int blockSize, int byteOffset) throws FileNotFoundException {
        this.file = new RandomAccessFile(file, "rw");
        this.blockSize = blockSize;
        this.byteOffset = byteOffset;
    }

    public FileBlockDeviceDriver(File file, int byteOffset) throws FileNotFoundException {
        this(file, 512, byteOffset);
    }

    public FileBlockDeviceDriver(File file) throws FileNotFoundException {
        this(file, 512, 0);
    }

    public FileBlockDeviceDriver(URL url, int blockSize, int byteOffset) throws IOException {
        this.byteOffset = byteOffset;
        ReadableByteChannel rbc = Channels.newChannel(url.openStream());
        File tempFile = File.createTempFile("libaums_file_blockdevice", ".bin");
        tempFile.deleteOnExit();
        FileOutputStream fos = new FileOutputStream(tempFile);
        fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);

        this.file = new RandomAccessFile(tempFile, "rw");
        this.blockSize = blockSize;
    }

    public FileBlockDeviceDriver(URL url, int byteOffset) throws IOException {
        this(url, 512, byteOffset);
    }

    public FileBlockDeviceDriver(URL url) throws IOException {
        this(url, 512, 0);
    }

    @Override
    public void init() throws IOException {

    }

    @Override
    public void read(long deviceOffset, ByteBuffer buffer) throws IOException {
        file.seek(deviceOffset * blockSize + byteOffset);
        int read = file.read(buffer.array(), buffer.position(), buffer.remaining());
        buffer.position(buffer.position() + read);
    }

    @Override
    public void write(long deviceOffset, ByteBuffer buffer) throws IOException {
        file.seek(deviceOffset * blockSize + byteOffset);
        file.write(buffer.array(), buffer.position(), buffer.remaining());
        buffer.position(buffer.limit());
    }

    @Override
    public int getBlockSize() {
        return blockSize;
    }
}

File: libaums/src/main/java/com/github/mjdev/libaums/driver/file/FileBlockDeviceDriver.kt
package com.github.mjdev.libaums.driver.file

import com.github.mjdev.libaums.driver.BlockDeviceDriver
import java.io.*
import java.net.URL
import java.nio.ByteBuffer
import java.nio.channels.Channels

/**
 * Created by magnusja on 01/08/17.
 */

class FileBlockDeviceDriver : BlockDeviceDriver {
    private var file: RandomAccessFile
    override var blockSize: Int = 0
        private set
    private var byteOffset: Int = 0

    override val blocks: Long
        get() = file.length() / blockSize

    @Throws(FileNotFoundException::class)
    @JvmOverloads
    constructor(file: File, byteOffset: Int = 0, blockSize: Int = 512) {
        this.file = RandomAccessFile(file, "rw")
        this.blockSize = blockSize
        this.byteOffset = byteOffset
    }

    @Throws(IOException::class)
    @JvmOverloads
    constructor(url: URL, byteOffset: Int = 0, blockSize: Int = 512) {
        this.byteOffset = byteOffset
        val rbc = Channels.newChannel(url.openStream())
        val tempFile = File.createTempFile("libaums_file_blockdevice", ".bin")
        tempFile.deleteOnExit()
        val fos = FileOutputStream(tempFile)
        fos.channel.transferFrom(rbc, 0, java.lang.Long.MAX_VALUE)

        this.file = RandomAccessFile(tempFile, "rw")
        this.blockSize = blockSize
    }

    @Throws(IOException::class)
    override fun init() {

    }

    @Throws(IOException::class)
    override fun read(deviceOffset: Long, buffer: ByteBuffer) {
        file.seek(deviceOffset * blockSize + byteOffset)
        val read = file.read(buffer.array(), buffer.position(), buffer.remaining())

        if (read == -1) {
            throw IOException("EOF")
        }

        buffer.position(buffer.position() + read)
    }

    @Throws(IOException::class)
    override fun write(deviceOffset: Long, buffer: ByteBuffer) {
        file.seek(deviceOffset * blockSize + byteOffset)
        file.write(buffer.array(), buffer.position(), buffer.remaining())
        buffer.position(buffer.limit())
    }
}
